addons/udes_sale_stock/tests/test_sale_order.py

```diff
@@ -0,0 +1,108 @@
+        """
+        Picking = self.env["stock.picking"]
+
+        sale_order = self._create_sale_order(self.apple.id, 10)
+        sale_order.action_confirm()
+
+        all_created_pickings = Picking.search([])
+        for created_picking in all_created_pickings:
+            with self.subTest(created_picking=created_picking):
+                self.assertIn(created_picking, sale_order.picking_ids)
+
+    def test_all_pickings_created_by_sale_order_are_cancelled(self):
+        """
+        Create a sale order for a single product, then cancel the sale order. Check all the pickings created due to the sale order are cancelled.
+        """
+        Picking = self.env["stock.picking"]
+
+        sale_order = self._create_sale_order(self.apple.id, 10)
+        sale_order.action_confirm()
+
+        all_created_pickings = Picking.search([])
+        sale_order.action_cancel()
+
+        for picking in all_created_pickings:
+            with self.subTest(picking=picking):
+                self.assertEqual(picking.state, "cancel")
+
+    def test_only_moves_created_by_sale_order_are_cancelled(self):
+        """
+        Create multiple sale orders for a single product, mimic refactoring by merging the pickings.
+        Then cancel one of the sale orders and check that then entire merged picking is not cancelled,
+        only the moves associated with the cancelled sale order are cancelled.
+        """
+        Picking = self.env["stock.picking"]
+        Move = self.env["stock.move"]
+
+        first_sale_order = self._create_sale_order(self.apple.id, 10)
+        first_sale_order.action_confirm()
+        first_picking = Picking.search([("picking_type_id", "=", self.picking_type_pick.id)])
+        first_move = first_picking.move_lines
+
+        second_sale_order = self._create_sale_order(self.apple.id, 5)
+        second_sale_order.action_confirm()
+        second_picking = Picking.search(
+            [("picking_type_id", "=", self.picking_type_pick.id), ("id", "!=", first_picking.id)]
+        )
+        second_move = Move.search(
+            [("picking_type_id", "=", self.picking_type_pick.id), ("id", "!=", first_move.id)]
+        )
+
+        created_picks = Picking.search([("picking_type_id", "=", self.picking_type_pick.id)])
+        self.assertEqual(len(created_picks), 2)
+        second_picking.move_lines.write({"picking_id": first_picking.id})
+        self.assertEqual(len(first_picking.move_lines), 2)
+
+        second_sale_order.action_cancel()
+        self.assertEqual(second_move.state, "cancel")
+    
+    def _create_sale_order(self, product_id, product_uom_qty):
+        """
+        Create a sale order for a the product_id
+        """
+        Partner = self.env["res.partner"]
+        SaleOrder = self.env["sale.order"]
+        SaleOrderLine = self.env["sale.order.line"]
+
+        partner = Partner.search([("name", "=", "Administrator")])
```

Comment:
Suggestion: Use the XML reference rather than search, unlikely but this could break if a test added another partner called "Administrator".

`partner = self.env.ref("base.partner_admin")` should do it.